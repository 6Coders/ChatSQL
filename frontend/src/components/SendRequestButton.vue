<template>
    <button type="button" @click="submitForm" v-if="!isSending">Invia</button>
    <button type="button" title="Interrompi" @click="stopSending" v-else>Interrompi</button>
</template>
  
  <script>
  import { ref } from 'vue'
  import axios from 'axios'
  export default {
    name: 'SendRequestButton',
    props: ['requestMessage'],
    setup(props, { emit }) {
      const isSending = ref(false)
      const CancelToken = axios.CancelToken
      let cancel

      async function generatePrompt() {
        return new Promise((resolve, reject) => {
          axios.post('/generateprompt', { userRequest: props.requestMessage }, {
            cancelToken: new CancelToken(function executor(c) {
              cancel = c
            })
          })
            .then(response => {
              resolve(response.data.result);
            })
            .catch(error => {
              if (axios.isCancel(error)) {
                reject('Request canceled')
              } else {
                reject(error)
              }
            })
        })
      }
  
      async function submitForm() {
        isSending.value = true
        try {
          const responseMessage = await generatePrompt()
          emit('submit', responseMessage)
        } catch (error) {
            emit('submit', 'Stopped')
            console.error(error)
        } finally {
          isSending.value = false
        }
      }
  
      const stopSending = () => {
        if (cancel) {
          cancel('Request canceled')
        }
      }
  
      return {
        submitForm,
        stopSending,
        isSending
      }
    }
  }
  </script>