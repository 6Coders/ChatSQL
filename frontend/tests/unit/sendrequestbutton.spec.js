import { shallowMount } from '@vue/test-utils'
import axios from 'axios'
import SendRequestButton from '@/components/SendRequestButton.vue'

jest.mock('axios')

describe('SendRequestButton.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SendRequestButton, {
      propsData: {
        requestMessage: 'test message'
      }
    })
  })

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('sends request when "Invia" button is clicked', async () => {
    const mockResponse = { data: { result: 'test result' } }
    axios.post.mockResolvedValue(mockResponse)

    await wrapper.find('button').trigger('click')

    expect(axios.post).toHaveBeenCalledWith('/generateprompt', { userRequest: 'test message' }, expect.anything())
    expect(wrapper.emitted().submit[0]).toEqual(['test result'])
  })

  it('stops sending when "Interrompi" button is clicked', async () => {
    const mockCancel = jest.fn()
    axios.CancelToken = jest.fn(() => ({
      token: {},
      cancel: mockCancel
    }))

    await wrapper.find('button').trigger('click') // Click "Invia" button
    await wrapper.find('button[title="Interrompi"]').trigger('click') // Click "Interrompi" button

    expect(mockCancel).toHaveBeenCalled()
  })
})