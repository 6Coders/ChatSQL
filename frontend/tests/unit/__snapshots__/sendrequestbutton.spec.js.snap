// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendRequestButton.vue renders correctly 1`] = `
<button
  type="button"
>
  Invia
</button>
`;
